src/QueryBuilder/QueryViewer/QueryViewer.js

```diff
@@ -0,0 +1,124 @@
+  accordionHeadline,
+  isInProgress,
+  inProgressTitle,
+  height,
+}) => {
+  const { changePage, pagination } = usePagination({ limit, offset });
+  const {
+    columnMapping,
+    defaultColumns,
+    defaultVisibleColumns,
+  } = useMemo(() => getTableMetadata(entityType), [entityType]);
+
+  useEffect(() => {
+    onPageChange?.(pagination);
+  }, [pagination]);
+
+  // set visible by default columns once
+  useEffect(() => {
+    if (defaultVisibleColumns && defaultColumns) {
+      onSetDefaultColumns?.(defaultColumns);
+      onSetDefaultVisibleColumns?.(defaultVisibleColumns);
+    }
+  }, [defaultVisibleColumns, defaultColumns]);
+
+  const renderTable = () => (
+    <>
+      <Row between="xs">
+        <Col xs={6}>
+          <Headline size="large" margin="none" tag="h3">
+            {headline}
+          </Headline>
+        </Col>
+        {headlineEnd}
+      </Row>
+      <Row>
+        <Col xs={12}>
+          <MultiColumnList
+            contentData={contentData}
+            columnMapping={columnMapping}
+            visibleColumns={visibleColumns}
+            pagingType={null}
+            onNeedMoreData={changePage}
+            height={height}
+          />
+          {showPagination && (
+            <PrevNextPagination
+              {...pagination}
+              totalCount={contentData?.length}
+              disabled={contentData?.length <= limit}
+              onChange={changePage}
+            />
+          )}
+        </Col>
+      </Row>
+    </>
+  );
+
+  const renderWithAccordion = () => (
+    <AccordionSet>
+      <Accordion
+        id="accordion-id"
+        label={accordionHeadline}
+      >
+        {renderTable()}
+      </Accordion>
+    </AccordionSet>
+  );
+
+  if (isInProgress) return <QueryLoader title={inProgressTitle} />;
+
+  return accordionHeadline ? renderWithAccordion() : renderTable();
+};
+
+QueryViewer.propTypes = {
+  accordionHeadline: PropTypes.string,
+  headline: PropTypes.oneOfType([PropTypes.string, PropTypes.element]),
+  headlineEnd: PropTypes.oneOfType([PropTypes.string, PropTypes.element]),
+  contentData: PropTypes.arrayOf(PropTypes.object),
+  entityType: PropTypes.shape({
+    id: PropTypes.string,
+    name: PropTypes.string,
+    labelAlias: PropTypes.string,
+    columns: PropTypes.arrayOf(PropTypes.shape({
+      name: PropTypes.string,
+      dataType: PropTypes.shape({
+        dataType: PropTypes.string,
+      }),
+      labelAlias: PropTypes.string,
+      visibleByDefault: PropTypes.bool,
+    })),
+  }),
+  visibleColumns: PropTypes.arrayOf(PropTypes.string),
+  onSetDefaultVisibleColumns: PropTypes.func,
+  onSetDefaultColumns: PropTypes.func,
+  limit: PropTypes.number,
+  offset: PropTypes.number,
+  onPageChange: PropTypes.func,
+  isInProgress: PropTypes.bool,
+  inProgressTitle: PropTypes.string,
+  height: PropTypes.number,
+  hidePagination: PropTypes.number,
+  showPagination: PropTypes.bool,
+};
+
+export default memo(QueryViewer);
```

Comment:
do you really need memo?

Reply:
and please use named exports

Reply:
I hope yes, we need memo for preventing situation of reculculating columns all the time if it will came from the same API as contentData. Will add named exports here

Reply:
gold rule: don't do any optimization until they are required

Reply:
Removed. But still feel that in this case the coast of memoizing less than benefits of recalculation in some cases. But we don't have big number of columns, and let it be.